keyshare.go

```diff
@@ -15,15 +15,20 @@ func NewKeyshareSecret() (*big.Int, error) {
 
 // NewKeyshareCommitments generates commitments for the keyshare server for given set of keys
 func NewKeyshareCommitments(secret *big.Int, keys []*gabikeys.PublicKey) (*big.Int, []*ProofPCommitment, error) {
-	// Generate randomizer value.
-	// Given that with this zero knowledge proof we are hiding a secret of length params[1024].Lm,
-	// normally we would use params[1024].LmCommit here. Generally LmCommit = Lm + Lh + Lstatzk,
-	// where Lstatzk is the level of security with which the proof hides the secret.
-	// However, params[1024].Lstatzk = 80 while everywhere else we use Lstatzk = 128.
-	// So instead of using params[1024].LmCommit we recompute it with the Lstatzk of key length 2048.
-	randLength := gabikeys.DefaultSystemParameters[1024].Lm +
-		gabikeys.DefaultSystemParameters[1024].Lh +
-		gabikeys.DefaultSystemParameters[2048].Lstatzk
+	// Generate randomizer value, whose length is specified by the LmCommit parameter.
+	// Generally LmCommit = Lm + Lh + Lstatzk, where Lstatzk is the level of security with which the
+	// proof hides the secret. Generally Lstatzk = 128, but for 1024 bit keys, Lstatzk = 80.
+	// So we prefer params[2048].LmCommit here, but if one of the keys is 1024 bits, then we have
+	// to fall back to params[1024].LmCommit, because otherwise the larger Lstatzk will cause
+	// the zero-knowledge proof response of the secret key to be too large, so that verification
+	// will fail (in ProofD.correctResponseSizes()).
+	randLength := gabikeys.DefaultSystemParameters[2048].LmCommit
+	for _, key := range keys {
+		if key.N.BitLen() == 1024 {
+			randLength = gabikeys.DefaultSystemParameters[1024].LmCommit
```

Comment:
I understand that the randomizer must be big enough to hide the actual secret. This works with the 1024 bit system parameter LmCommit since the secret size itself is based on the 1024 bit system parameter Lm. So the fix is fine. However, it bothers me that we don't additionally check the secrets length here. Currently the secret's length is hardcoded in `NewKeyshareSecret` and if we should ever change this to the 4k default, issuance with a key bitlength of 1k should for sure not be possible. Still, I feel an additional check that the length of the secret does not exceed `DefaultSystemParameters[1024].Lm` before using the `DefaultSystemParameters[1024].LmCommit` would add value.